bliss/simulator/background.py

```diff
@@ -1,48 +1,64 @@
 
 
 class ConstantBackground(nn.Module):
-    def __init__(self, background: Tuple[float, ...]):
+    def __init__(self, background: Tuple[float, ...], **_kwargs):
         super().__init__()
-        bg: Tensor = torch.tensor(background)
-        bg = rearrange(bg, "c -> 1 c 1 1")
+        bg: Tensor = torch.tensor([background])
+        bg = rearrange(bg[0], "c -> 1 c 1 1")
         self.register_buffer("background", bg, persistent=False)
 
-    def sample(self, shape) -> Tensor:
+    def sample(self, shape, **_kwargs) -> Tensor:
         assert isinstance(self.background, Tensor)
         batch_size, c, hlen, wlen = shape
         return self.background.expand(batch_size, c, hlen, wlen)
 
 
 class SimulatedSDSSBackground(nn.Module):
-    def __init__(self, sdss_dir, run, camcol, field, bands):
+    def __init__(self, sdss_fields: DictConfig):
         super().__init__()
-        sdss_data = SloanDigitalSkySurvey(
-            sdss_dir=sdss_dir,
-            run=run,
-            camcol=camcol,
-            fields=(field,),
-            bands=bands,
-        )
-        background = torch.from_numpy(sdss_data[0]["background"])
-        background = rearrange(background, "c h w -> 1 c h w", c=len(bands))
+        sdss_dir = sdss_fields["dir"]
+        field_list = sdss_fields["field_list"]
+        bands = sdss_fields["bands"]  # use same bands across fields
+
+        # Add all backgrounds from specified run/col/field to list
+        backgrounds = []
+        for param_obj in field_list:
+            params: Dict = OmegaConf.to_container(param_obj)  # type: ignore
+            sdss = SloanDigitalSkySurvey(sdss_dir=sdss_dir, bands=bands, **params)
+            backgrounds.extend([field["background"] for field in sdss])  # type: ignore
+
+        background = torch.from_numpy(np.stack(backgrounds, axis=0))
+        if len(background.size()) == 3:  # convert to 4d tensor if only one background
+            background = rearrange(background, "c h w -> 1 c h w")
+
         self.register_buffer("background", background, persistent=False)
-        self.height, self.width = self.background.shape[-2:]
+        self.height, self.width = self.background.shape[-2:]  # type: ignore
 
-    def sample(self, shape) -> Tensor:
+    def sample(self, shape, rcf_indices=None) -> Tensor:
         assert isinstance(self.background, Tensor)
         batch_size, c, hlen, wlen = shape
         assert self.background.shape[1] == c
+
+        # select region to sample from (same for all images in batch for simplicity)
         h_diff, w_diff = self.height - hlen, self.width - wlen
         h = 0 if h_diff == 0 else np.random.randint(h_diff)
         w = 0 if w_diff == 0 else np.random.randint(w_diff)
-        bg = self.background[:, :, h : (h + hlen), w : (w + wlen)]
-        return bg.expand(batch_size, -1, -1, -1)
+
+        # sample region from random background for each image in batch
+        if rcf_indices is not None:
```

Comment:
Any way to get rid of this conditional statement? i.e., do we definitely need both cases? Code with extra branching is a bit harder to test and for other developers to understand

Reply:
Since we were unsure last week of whether we should choose backgrounds and psfs independently or specify which R/C/F to use for both, I left it as an option in case we change our mind later while still finalizing the model. Right now it's set up to always use the RCF - should I remove the option completely, or would it be better to just document the function better?

Reply:
Probably better if pick one way

Reply:
Got it, I'll make it always require the RCF then. Maybe when we restructure it to take catalog-agnostic input we can add back the option of sampling independently